EventFilter/L1TRawToDigi/plugins/implementations_stage2/ZDCUnpacker.cc

```diff
@@ -0,0 +1,71 @@
+#include "FWCore/MessageLogger/interface/MessageLogger.h"
+#include "EventFilter/L1TRawToDigi/plugins/UnpackerFactory.h"
+
+#include "L1Trigger/L1TCalorimeter/interface/CaloTools.h"
+
+#include "L1TObjectCollections.h"
+
+#include "L1TStage2Layer2Constants.h"
+#include "ZDCUnpacker.h"
+
+namespace l1t {
+  namespace stage2 {
+    ZDCUnpacker::ZDCUnpacker() : EtSumZDCCopy_(0) {}
+
+    bool ZDCUnpacker::unpack(const Block& block, UnpackerCollections* coll) {
+      using namespace l1t::stage2;
+
+      LogDebug("L1T") << "Block ID  = " << block.header().getID() << " size = " << block.header().getSize();
+
+      int nBX = int(
+          ceil(block.header().getSize() / layer2::demux::nOutputFramePerBX));  // Since there 6 frames per demux output event
+      // expect the first four frames to be the first 4 EtSum objects reported per event (see CMS IN-2013/005)
+
+      // Find the central, first and last BXs
+      int firstBX = -(ceil((double)nBX / 2.) - 1);
+      int lastBX;
+      if (nBX % 2 == 0) {
+        lastBX = ceil((double)nBX / 2.);
+      } else {
+        lastBX = ceil((double)nBX / 2.) - 1;
+      }
+
+      auto res_ = static_cast<L1TObjectCollections*>(coll)->getZDCSums(EtSumZDCCopy_);
+      res_->setBXRange(firstBX, lastBX);
+
+      LogDebug("L1T") << "nBX = " << nBX << " first BX = " << firstBX << " lastBX = " << lastBX;
+
+      // Loop over multiple BX and fill EtSums collection
+      for (int bx = firstBX; bx <= lastBX; bx++) {
+        // ZDC -
+        int iFrame = (bx - firstBX) * layer2::demux::nOutputFramePerBX;
+
+        uint32_t raw_data = block.payload().at(iFrame +1); // ZDC - info is found on frame 1 of each bx
+        
+        l1t::EtSum zdcm{l1t::EtSum::kZDCM};
+        zdcm.setHwPt(raw_data & 0xFFFF);
+        zdcm.setP4(l1t::CaloTools::p4Demux(&zdcm));
+
+        LogDebug("L1T") << "ZDC -: pT " << zdcm.hwPt() << " bx " << bx;
+
+        res_->push_back(bx, zdcm);
+        

```

Comment:
I think the EtSums are selected using the sumType Enum (rather than index in the EtSum vector):
https://github.com/cms-sw/cmssw/blob/master/L1Trigger/L1TGlobal/src/EnergySumZdcCondition.cc#L139
https://github.com/cms-sw/cmssw/blob/master/L1Trigger/L1TGlobal/src/EnergySumCondition.cc#L125